Add ConPty APIs

davidhewitt opened the pull request:
Closes #695.

I added the bare minimum to cover those three functions. These APIs will only work on the very newest Windows. Does anything need to be done to feature gate them?

davidhewitt committed:
Add ConPty APIs

retep998 commented:
There's no need to feature gate things based on Windows version because extern functions are only linked to when used. That said, if you *do* use these functions then your program won't run at all on older Windows versions unless you use `GetProcAddress` to obtain pointers to the functions instead of using these extern functions directly.

davidhewitt commented:
Cool. I'd rather not use `GetProcAddress` but it's going to be quite a small segment of windows users able to use these currently! 😄 

Do you need anything else to make this PR mergeable?

chyyran commented:
1809 is now GA, so ConPTY should be available on updated Windows 10 systems.

retep998 commented on src/um/wincon.rs:
```diff
@@ -538,3 +538,4 @@ extern "system" {
         ExeNameBufferLength: DWORD,
     ) -> DWORD;
 }
+pub type HPCON = *mut c_void;
```
This is actually defined in `wincontypes.h`

davidhewitt replied:
I can move it to a new file "wincontypes.rs". Do I need to add an accompanying feature for this new file?

I'd originally not done so because it seemed a bit overkill to make a new file for just the one line! While I'm at it want me to attempt to wrap whatever else is defined in `wincontypes.h`?

retep998 replied:
You would need to add a feature for the new file. It's up to you whether you want to bind whatever else is in there.

davidhewitt replied:
Great - will see how much I can find time for this weekend.

davidhewitt committed:
Move definitions from wincon to wincontypes

davidhewitt commented:
I've moved HPCON to `wincontypes.rs` and introduced a new feature as discussed. Turns out a number of the struct definitions in `wincon.rs` are also defined in `wincontypes.h`, so I've moved those at the same time.

retep998 commented:
When moving existing definitions to a different header, please put in `pub use` statements so that code still relying on the definitions in the old location is not broken.

Also when adding a new header, you also have to update the feature array in `build.rs`.

davidhewitt commented:
Thanks - will fixup tomorrow morning :)

davidhewitt committed:
Fix issues with wincontypes move

davidhewitt committed:
Final correction to build.rs for wincontypes

davidhewitt commented:
Have corrected all that now. Many thanks for your patience in helping refine all the bits I originally missed. Would a separate PR be useful to extend the notes in CONTRIBUTING to make those points clearer?

davidhewitt commented:
Just wondering if there's anything else you'd like to see to have this merged?

davidhewitt commented:
@retep998 just pinging to say we're hoping to merge https://github.com/jwilm/alacritty/pull/1762 but would like to see these definitions merged into winapi first (we're using this patch I'm providing). Please let me know if there's anything left I can do to help this PR merge.

davidhewitt commented:
Thanks!
